### PR TITLE
Enable tracer logging for BQ ReadRows in V1

### PR DESCRIPTION
**Pull request by nileyadav**

### CODE REVIEW

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/direct/Scala213BigQueryRDD.java [hunk 1]
```diff
@@ -83,6 +97,7 @@ public scala.collection.Iterator<InternalRow> compute(Partition split, TaskConte
             bigQueryClientFactory,
             request,
             options.toReadSessionCreatorConfig().toReadRowsHelperOptions());
+    tracer.startStream();
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/direct/Scala213BigQueryRDD.java:100]: Should be in `ReadRowsHelper.readRows()`

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
1. If the BigQueryStorageReadRowsTracer field is optional (may get null) please replace its occurrences with Optional<BigQueryStorageReadRowsTracer>
2. `tracer.startStream()` Should be in ReadRowsHelper.readRows()

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java [hunk 2]
```diff
@@ -50,7 +52,8 @@ public ArrowBinaryIterator(
       List<String> columnsInOrder,
       ByteString schema,
       ByteString rowsInBytes,
-      Optional<StructType> userProvidedSchema) {
+      Optional<StructType> userProvidedSchema,
+      BigQueryStorageReadRowsTracer bigQueryStorageReadRowsTracer) {
     BufferAllocator allocator =
         ArrowUtil.newRootAllocator(maxAllocation)
             .newChildAllocator("ArrowBinaryIterator", 0, maxAllocation);
```
[the remark this answers is not in the document]
Comment (nileyadav) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java:59]: Not relevant to change in review, won't fix.

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryUtil.java [hunk 3]
```diff
@@ -240,6 +242,20 @@ public static boolean schemaEquals(
     return fieldListEquals(s1.getFields(), s2.getFields(), enableModeCheckForSchemaFields);
   }
 
+  /**
+   * Create a list of read stream names of given read session.
+   *
+   * @param readSession
```
[the remark this answers is not in the document]
Comment (nileyadav) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryUtil.java]: Fixed

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java [hunk 4]
```diff
@@ -104,6 +109,9 @@ private Iterator<InternalRow> toArrowRows(VectorSchemaRoot root, List<String> na
 
     ColumnarBatch batch = new ColumnarBatch(columns);
     batch.setNumRows(root.getRowCount());
+    if (bigQueryStorageReadRowsTracer != null) {
```
[the remark this answers is not in the document]
Comment (nileyadav) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java]: Made the argument optional.

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/direct/Scala213BigQueryRDD.java (hunk 1 above)
Comment (nileyadav) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/direct/Scala213BigQueryRDD.java]: Fixed

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java [hunk 5]
```diff
@@ -50,7 +52,8 @@ public ArrowBinaryIterator(
       List<String> columnsInOrder,
       ByteString schema,
       ByteString rowsInBytes,
-      Optional<StructType> userProvidedSchema) {
+      Optional<StructType> userProvidedSchema,
+      Optional<BigQueryStorageReadRowsTracer> bigQueryStorageReadRowsTracer) {
```
Comment (emkornfield) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java:56]: unused?

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (emkornfield) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java:59]: this seems like it is probably an issue in general, though lets make sure we track it someplace?

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java [hunk 6]
```diff
@@ -53,17 +56,23 @@ public AvroBinaryIterator(
...
     this.userProvidedSchema = userProvidedSchema;
+    this.bigQueryStorageReadRowsTracer = bigQueryStorageReadRowsTracer;
   }
 
   @Override
   public boolean hasNext() {
     try {
+      // Avro iterator is used in both V1 and V2
+      if (bigQueryStorageReadRowsTracer.isPresent() && in.isEnd()) {
+        bigQueryStorageReadRowsTracer.get().rowsParseFinished(numberOfRowsParsed);
```
Comment (emkornfield) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java:74]: i think this potentially mingles parsing time with application time.

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java [hunk 7]
```diff
@@ -73,6 +82,11 @@ public boolean hasNext() {
   @Override
   public InternalRow next() {
     try {
+      // TODO(nileshny) : confirm if this is correct way to get number of Avro rows parsed
+      try {
+        numberOfRowsParsed = Math.addExact(numberOfRowsParsed, 1);
```
Comment (emkornfield) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java:87]: why Math.addExact?

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java (hunk 5 above)
Comment (nileyadav) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java:56]: rowParseFinished is called on tracer on line 113

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java (hunk 7 above)
Comment (nileyadav) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java:87]: to detect overflow

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java (hunk 6 above)
Comment (nileyadav) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java:74]: @emkornfield  I see. Which is right place for "rowParseFinished?
Is it after "SchemaConverters.convertToInternalRow" in next() method? Will that cover parsing time for Avro?

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java [hunk 8]
```diff
@@ -104,6 +109,9 @@ private Iterator<InternalRow> toArrowRows(VectorSchemaRoot root, List<String> na
 
     ColumnarBatch batch = new ColumnarBatch(columns);
     batch.setNumRows(root.getRowCount());
+    if (bigQueryStorageReadRowsTracer.isPresent()) {
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java:112]: Cleaner way:
```
bigQueryStorageReadRowsTracer.ifPresent(tracer -> tracer.rowsParseFinished(root.getRowCount()));
```

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java [hunk 9]
```diff
@@ -73,6 +82,11 @@ public boolean hasNext() {
...
+      } catch (ArithmeticException ae) {;
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java:88]: Empty catch - either throw or explain why it is ignored

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java (hunk 8 above)
Comment (nileyadav) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java:112]: Done

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java (hunk 7 above)
Comment (nileyadav) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/AvroBinaryIterator.java:87]: I replaced int with long as that's appropriate type in this case.

**Comment by davidrabinowitz** [comment]
@nileyadav the tracer needs to be serializable
```
Step #5 - "integration-tests-2.4": [ERROR]   Spark24ReadIntegrationTest>ReadIntegrationTestBase.testArrowCompressionCodec:372 » Spark Job aborted due to stage failure: Failed to serialize task 22, not attempting to retry it. Exception during serialization: java.io.NotSerializableException: java.util.Optional
Step #5 - "integration-tests-2.4": Serialization stack:
Step #5 - "integration-tests-2.4": 	- object not serializable (class: java.util.Optional, value: Optional.empty)
Step #5 - "integration-tests-2.4": 	- field (class: com.google.cloud.spark.bigquery.ReadRowsResponseToInternalRowIteratorConverter$Avro, name: bigQueryStorageReadRowsTracer, type: class java.util.Optional)
Step #5 - "integration-tests-2.4": 	- object (class com.google.cloud.spark.bigquery.ReadRowsResponseToInternalRowIteratorConverter$Avro, com.google.cloud.spark.bigquery.ReadRowsResponseToInternalRowIteratorConverter$Avro@100a6cc9)
Step #5 - "integration-tests-2.4": 	- field (class: com.google.cloud.spark.bigquery.v2.context.BigQueryInputPartitionContext, name: converter, type: interface com.google.cloud.spark.bigquery.ReadRowsResponseToInternalRowIteratorConverter)
Step #5 - "integration-tests-2.4": 	- object (class com.google.cloud.spark.bigquery.v2.context.BigQueryInputPartitionContext, com.google.cloud.spark.bigquery.v2.context.BigQueryInputPartitionContext@23824eba)
Step #5 - "integration-tests-2.4": 	- field (class: com.google.cloud.spark.bigquery.v2.Spark24InputPartition, name: context, type: interface com.google.cloud.spark.bigquery.v2.context.InputPartitionContext)
Step #5 - "integration-tests-2.4": 	- object (class com.google.cloud.spark.bigquery.v2.Spark24InputPartition, com.google.cloud.spark.bigquery.v2.Spark24InputPartition@12d09eca)
Step #5 - "integration-tests-2.4": 	- field (class: org.apache.spark.sql.execution.datasources.v2.DataSourceRDDPartition, name: inputPartition, type: interface org.apache.spark.sql.sources.v2.reader.InputPartition)
Step #5 - "integration-tests-2.4": 	- object (class org.apache.spark.sql.execution.datasources.v2.DataSourceRDDPartition, org.apache.spark.sql.execution.datasources.v2.DataSourceRDDPartition@0)
Step #5 - "integration-tests-2.4": 	- field (class: org.apache.spark.scheduler.ResultTask, name: partition, type: interface org.apache.spark.Partition)
Step #5 - "integration-tests-2.4": 	- object (class org.apache.spark.scheduler.ResultTask, ResultTask(3, 0))
Step #5 - "integration-tests-2.4": [ERROR]   Spark24ReadIntegrationTest>ReadIntegrationTestBase.testOrAcrossColumnsAndFormats:347 » Spark Job aborted due to stage failure: Failed to serialize task 194, not attempting to retry it. Exception during serialization: java.io.NotSerializableException: java.util.Optional
Step #5 - "integration-tests-2.4": Serialization stack:
Step #5 - "integration-tests-2.4": 	- object not serializable (class: java.util.Optional, value: Optional.empty)
Step #5 - "integration-tests-2.4": 	- field (class: com.google.cloud.spark.bigquery.ReadRowsResponseToInternalRowIteratorConverter$Avro, name: bigQueryStorageReadRowsTracer, type: class java.util.Optional)
Step #5 - "integration-tests-2.4": 	- object (class com.google.cloud.spark.bigquery.ReadRowsResponseToInternalRowIteratorConverter$Avro, com.google.cloud.spark.bigquery.ReadRowsResponseToInternalRowIteratorConverter$Avro@a02351f)
Step #5 - "integration-tests-2.4": 	- field (class: com.google.cloud.spark.bigquery.v2.context.BigQueryInputPartitionContext, name: converter, type: interface com.google.cloud.spark.bigquery.ReadRowsResponseToInternalRowIteratorConverter)
Step #5 - "integration-tests-2.4": 	- object (class com.google.cloud.spark.bigquery.v2.context.BigQueryInputPartitionContext, com.google.cloud.spark.bigquery.v2.context.BigQueryInputPartitionContext@aab5fea)
Step #5 - "integration-tests-2.4": 	- field (class: com.google.cloud.spark.bigquery.v2.Spark24InputPartition, name: context, type: interface com.google.cloud.spark.bigquery.v2.context.InputPartitionContext)
Step #5 - "integration-tests-2.4": 	- object (class com.google.cloud.spark.bigquery.v2.Spark24InputPartition, com.google.cloud.spark.bigquery.v2.Spark24InputPartition@12b99bc6)
Step #5 - "integration-tests-2.4": 	- field (class: org.apache.spark.sql.execution.datasources.v2.DataSourceRDDPartition, name: inputPartition, type: interface org.apache.spark.sql.sources.v2.reader.InputPartition)
Step #5 - "integration-tests-2.4": 	- object (class org.apache.spark.sql.execution.datasources.v2.DataSourceRDDPartition, org.apache.spark.sql.execution.datasources.v2.DataSourceRDDPartition@0)
Step #5 - "integration-tests-2.4": 	- field (class: org.apache.spark.scheduler.ResultTask, name: partition, type: interface org.apache.spark.Partition)
Step #5 - "integration-tests-2.4": 	- object (class org.apache.spark.scheduler.ResultTask, ResultTask(38, 0))
```

**Pull request merged**